File: android/app/src/main/java/com/dibay/app/call/NativeActiveCallHeartbeatOwner.java
package com.dibay.app.call;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Session-bound native active-call heartbeat.
 *
 * Production liveness for native-established calls. This intentionally mirrors the web
 * active-call heartbeat cadence, not the 300s shadow presence lease cadence.
 */
public final class NativeActiveCallHeartbeatOwner {
  private static final String TAG = "DibayNativeActiveHeartbeat";

  /** Mirror lib/call/native/call-heartbeat-watchdog.ts CALL_HEARTBEAT_INTERVAL_MS. */
  public static final long ACTIVE_HEARTBEAT_INTERVAL_MS = 10_000L;

  public interface HeartbeatCallback {
    void onDone(boolean ok, int status, String error);
  }

  public interface HeartbeatTransport {
    void heartbeat(Context app, String callId, HeartbeatCallback callback);
  }

  public interface LiveGate {
    boolean isHeartbeatLive(String callId);
  }

  private static final Object LOCK = new Object();
  private static final Handler MAIN = new Handler(Looper.getMainLooper());
  private static final AtomicInteger GENERATION = new AtomicInteger(0);

  private static String activeCallId = null;
  private static int activeGeneration = 0;
  private static Runnable pending = null;
  private static HeartbeatTransport transport = null;
  private static LiveGate liveGate = null;
  private static int heartbeatCountForTests = 0;

  private NativeActiveCallHeartbeatOwner() {}

  public static void start(
      Context context, String callId, HeartbeatTransport heartbeatTransport, LiveGate gate) {
    if (context == null || callId == null || callId.trim().isEmpty()) return;
    if (heartbeatTransport == null || gate == null) return;
    Context app = context.getApplicationContext();
    String sid = callId.trim();
    int gen;
    synchronized (LOCK) {
      clearPendingLocked();
      activeCallId = sid;
      activeGeneration = GENERATION.incrementAndGet();
      gen = activeGeneration;
      transport = heartbeatTransport;
      liveGate = gate;
      heartbeatCountForTests = 0;
    }
    Log.i(TAG, "active_heartbeat_start callId=" + sid + " intervalMs=" + ACTIVE_HEARTBEAT_INTERVAL_MS);
    fireHeartbeat(app, sid, gen, "immediate_connected");
  }

  public static void stop(String callId, String reason) {
    String sid = callId != null ? callId.trim() : "";
    synchronized (LOCK) {
      if (sid.isEmpty() || activeCallId == null || !sid.equals(activeCallId)) return;
      Log.i(TAG, "active_heartbeat_stop callId=" + sid + " reason=" + safe(reason));
      clearPendingLocked();
      activeCallId = null;
      activeGeneration = GENERATION.incrementAndGet();
      transport = null;
      liveGate = null;
    }
  }

  public static boolean isActiveForTests(String callId) {
    synchronized (LOCK) {
      return callId != null && callId.trim().equals(activeCallId);
    }
  }

  public static int heartbeatCountForTests() {
    synchronized (LOCK) {
      return heartbeatCountForTests;
    }
  }

  public static void resetForTests() {
    synchronized (LOCK) {
      clearPendingLocked();
      activeCallId = null;
      activeGeneration = GENERATION.incrementAndGet();
      transport = null;
      liveGate = null;
      heartbeatCountForTests = 0;
    }
  }

  public static void flushForTests() {
    Runnable next;
    synchronized (LOCK) {
      next = pending;
    }
    if (next != null) {
      MAIN.removeCallbacks(next);
      next.run();
    }
  }

  private static void fireHeartbeat(Context app, String sid, int gen, String source) {
    HeartbeatTransport t;
    LiveGate gate;
    synchronized (LOCK) {
      if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) return;
      t = transport;
      gate = liveGate;
    }
    if (t == null || gate == null) return;
    if (!gate.isHeartbeatLive(sid)) {
      stop(sid, "not_live_gate");
      return;
    }
    t.heartbeat(
        app,
        sid,
        (ok, status, error) -> {
          synchronized (LOCK) {
            if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) return;
            heartbeatCountForTests += 1;
          }
          if (!ok) {
            String err = error != null ? error : "";
            if ("not_live".equals(err) || err.contains("not_live")) {
              stop(sid, "server_not_live");
              return;
            }
            Log.w(TAG, "active_heartbeat_failed callId=" + sid + " err=" + err + " source=" + source);
          } else {
            Log.i(TAG, "active_heartbeat_ok callId=" + sid + " source=" + source);
          }
          scheduleNext(app, sid, gen);
        });
  }

  private static void scheduleNext(Context app, String sid, int gen) {
    synchronized (LOCK) {
      if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) return;
      clearPendingLocked();
      Runnable runnable =
          () -> {
            synchronized (LOCK) {
              if (activeCallId == null || !sid.equals(activeCallId) || gen != activeGeneration) return;
              pending = null;
            }
            fireHeartbeat(app, sid, gen, "periodic");
          };
      pending = runnable;
      MAIN.postDelayed(runnable, ACTIVE_HEARTBEAT_INTERVAL_MS);
    }
  }

  private static void clearPendingLocked() {
    if (pending != null) {
      MAIN.removeCallbacks(pending);
      pending = null;
    }
  }

  private static String safe(String value) {
    return value == null ? "" : value.trim();
  }
}

File: ios/App/App/Call/NativeActiveCallHeartbeatOwner.swift
import Foundation

/**
 * Session-bound native active-call heartbeat.
 *
 * Production liveness for native-established calls. Cadence mirrors the web active-call
 * heartbeat and is intentionally separate from the 300s shadow native presence lease.
 */
enum NativeActiveCallHeartbeatOwner {
  /// Mirror lib/call/native/call-heartbeat-watchdog.ts CALL_HEARTBEAT_INTERVAL_MS.
  static let activeHeartbeatIntervalMs: Int = 10_000

  typealias HeartbeatCallback = (_ ok: Bool, _ status: Int, _ error: String?) -> Void
  typealias HeartbeatTransport = (_ callId: String, _ completion: @escaping HeartbeatCallback) -> Void
  typealias LiveGate = (_ callId: String) -> Bool

  private static let lock = NSLock()
  private static var activeCallId: String?
  private static var activeGeneration: Int = 0
  private static var generationCounter: Int = 0
  private static var workItem: DispatchWorkItem?
  private static var transport: HeartbeatTransport?
  private static var liveGate: LiveGate?
  private static var heartbeatCountForTests: Int = 0
  private static let queue = DispatchQueue(label: "com.dibay.native.active-heartbeat")

  static func start(callId: String, heartbeatTransport: @escaping HeartbeatTransport, gate: @escaping LiveGate) {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !sid.isEmpty else { return }
    let gen: Int
    lock.lock()
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    gen = activeGeneration
    activeCallId = sid
    transport = heartbeatTransport
    liveGate = gate
    heartbeatCountForTests = 0
    lock.unlock()
    DibayCallLog.info(
      "active_heartbeat_start",
      sessionId: sid,
      detail: "intervalMs=\(activeHeartbeatIntervalMs)"
    )
    fireHeartbeat(sid: sid, gen: gen, source: "immediate_connected")
  }

  static func stop(callId: String, reason: String) {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    lock.lock()
    defer { lock.unlock() }
    guard !sid.isEmpty, activeCallId == sid else { return }
    DibayCallLog.info("active_heartbeat_stop", sessionId: sid, detail: "reason=\(reason)")
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    activeCallId = nil
    transport = nil
    liveGate = nil
  }

  static func isActiveForTests(_ callId: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return activeCallId == callId.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  static func heartbeatCountForTestsValue() -> Int {
    lock.lock()
    defer { lock.unlock() }
    return heartbeatCountForTests
  }

  static func resetForTests() {
    lock.lock()
    clearPendingLocked()
    generationCounter += 1
    activeGeneration = generationCounter
    activeCallId = nil
    transport = nil
    liveGate = nil
    heartbeatCountForTests = 0
    lock.unlock()
  }

  static func flushForTests() {
    lock.lock()
    let item = workItem
    workItem = nil
    lock.unlock()
    item?.cancel()
    item?.perform()
  }

  private static func fireHeartbeat(sid: String, gen: Int, source: String) {
    let t: HeartbeatTransport?
    let gate: LiveGate?
    lock.lock()
    guard activeCallId == sid, gen == activeGeneration else {
      lock.unlock()
      return
    }
    t = transport
    gate = liveGate
    lock.unlock()
    guard let t, let gate else { return }
    guard gate(sid) else {
      stop(callId: sid, reason: "not_live_gate")
      return
    }
    t(sid) { ok, status, error in
      lock.lock()
      guard activeCallId == sid, gen == activeGeneration else {
        lock.unlock()
        return
      }
      heartbeatCountForTests += 1
      lock.unlock()
      if !ok {
        let err = error ?? ""
        if err == "not_live" || err.contains("not_live") {
          stop(callId: sid, reason: "server_not_live")
          return
        }
        DibayCallLog.info(
          "active_heartbeat_failed",
          sessionId: sid,
          detail: "err=\(err) source=\(source) status=\(status)"
        )
      } else {
        DibayCallLog.info("active_heartbeat_ok", sessionId: sid, detail: "source=\(source)")
      }
      scheduleNext(sid: sid, gen: gen)
    }
  }

  private static func scheduleNext(sid: String, gen: Int) {
    lock.lock()
    guard activeCallId == sid, gen == activeGeneration else {
      lock.unlock()
      return
    }
    clearPendingLocked()
    let work = DispatchWorkItem {
      fireHeartbeat(sid: sid, gen: gen, source: "periodic")
    }
    workItem = work
    lock.unlock()
    queue.asyncAfter(deadline: .now() + .milliseconds(activeHeartbeatIntervalMs), execute: work)
  }

  private static func clearPendingLocked() {
    workItem?.cancel()
    workItem = nil
  }
}
